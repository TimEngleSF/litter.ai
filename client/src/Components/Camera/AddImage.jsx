import React, { useRef } from 'react';
import { Link } from 'react-router-dom';
import { CiCirclePlus } from 'react-icons/ci';

export default function AddImage({ setImage }) {
  const captureRef = useRef(null);

  const openCamera = () => {
    captureRef.current.click();
  };

  const handleImage = (e) => {
    const imageData = [e.target.files[0]];
    setImage({
      imagePreview: URL.createObjectURL(
        new Blob(imageData, { type: 'application/zip' }),
      ),
      imageFile: e.target.files[0],
    });
  };
  return (
    <>
      { /* Image has not been uploaded, opens device's camera or gives option to upload an image */}
      <div className="capture-top-section">
        <CiCirclePlus onClick={openCamera} className="capture-icon" />
        <input
          type="file"
          accept="image/*"
          id="cameraFileInput"
          ref={captureRef}
          capture="environment"
          onChange={handleImage}
          style={{ display: 'none' }}
        />
        <div className="capture-icon-bg" />
        <h1>Take a photo of the item</h1>
        <p>
          <span>Click on the icon to open up your device&#x27;s camera.</span>
          <br />
          <span>Please take the photo in portrait mode for best results.</span>
        </p>
      </div>
      <div className="capture-buttons lower-buttons">
        <Link to="/">
          <button className="button-home" type="button">
            Home
          </button>
        </Link>
      </div>
    </>
  );
}
