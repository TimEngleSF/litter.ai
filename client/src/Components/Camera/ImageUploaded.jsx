import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import Dropdown from '../Dropdown';

export default function ImageUploaded({
  image,
  setImage,
}) {
  const [imageSubmitted, setImageSubmitted] = useState(false);
  const [imageCategory, setImageCategory] = useState('glass');

  return (
    <div className="capture-image-wrapper">
      <div className="capture-image-preview">
        <button
          type="button"
          className="image-x-button"
          onClick={() => { setImage(null); }}
        >
          &#x2715;
        </button>
        <img alt="" src={image.imagePreview} className="capture-image" />
      </div>
      {/* Image has been uploaded and submitted to the AI */}
      {imageSubmitted
        ? (
          <div className="capture-classify">
            <h1>Help us classify</h1>
            <div className="capture-dropdown">
              <Dropdown
                setImageCategory={setImageCategory}
                imageCategory={imageCategory}
              />
            </div>
            <p>
              Is this&nbsp;
              {imageCategory}
              ?
              If it is not,
              please select the correct category from the dropdown.
            </p>
            <div className="capture-confirm-button lower-buttons">
              <Link to={`/success/${imageCategory}`}>
                <button type="button">
                  Confirm
                </button>
              </Link>
            </div>
          </div>
        )
        : ( // Image preview before sending to the AI
          <div className="capture-submit-button lower-buttons">
            <button
              type="button"
              onClick={() => { setImageSubmitted(true); }}
            >
              Submit
            </button>
          </div>
        )}
    </div>
  );
}
