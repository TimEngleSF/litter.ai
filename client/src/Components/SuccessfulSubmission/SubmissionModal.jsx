import React from 'react';
import parse from 'html-react-parser';

export default function SubmissionModal({
  modalOpen,
  setModalOpen,
  type,
}) {
  return (
    <div className="modal-container" data-testid="modal">
      <div className="modal-container-top">
        <h2>{type.name}</h2>
        <button
          type="button"
          className="modal-x-button"
          onClick={() => { setModalOpen(!modalOpen); }}
        >
          &#x2715;
        </button>
      </div>
      <hr />
      <div className="modal-text">{parse(type.desc_long)}</div>
      <button
        type="button"
        className="modal-got-it-button"
        data-testid="modal-got-it-button"
        onClick={() => { setModalOpen(!modalOpen); }}
      >
        Got it
      </button>
    </div>
  );
}
