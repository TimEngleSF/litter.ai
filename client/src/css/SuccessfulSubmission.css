.category-container {
  width: 100%;
  height: 100dvh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.category-container
.category-wrapper {
  height: 100dvh;
  width: 100dvw;
  max-width: 768px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
}

.category-container
.category-wrapper h4 {
  font-size: 1.5rem;
  padding-top: 4rem;
  margin-bottom: 0;
}

.category-container
.category-wrapper h1 {
  text-transform: capitalize;
}

.category-container
.category-wrapper
.category-icon {
  font-size: 10rem;
  color: var(--main-green);
}

.category-container
.category-wrapper
.category-short-desc {
  width: 85%;
}

.category-container
.category-wrapper
.category-short-desc p {
  font-size: 1.2rem;
  text-align: left;
  margin: 0.7rem 0;
}

.category-container
.category-modal {
  width: 35dvw;
  max-width: 100%;
  max-height: 45dvh;
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;
  z-index: 2;
}

/******** Modal ********/

.modal-container {
  min-height: 20dvh;
  max-height: 60dvh;
  background-color: var(--white);
  box-shadow: 0.75em 0.75em 1.75em -0.5em var(--shadow-color);
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border-radius: 1.5rem;
  z-index: 2;
}

.modal-container
.modal-container-top {
  width: 85%;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
}

.modal-container
.modal-container-top h2 {
  margin: 2rem 0 0 0;
  text-transform: capitalize;
}

.modal-container
.modal-x-button {
  border: none;
  background: none;
  font-size: 1rem;
  font-weight: bold;
  color: var(--black);
}

.modal-container
.modal-x-button:hover {
  cursor: pointer;
}

.modal-container hr {
  border: none;
  height: 1px;
  background-color: #000;
  margin-top: 1rem;
  width: 90%;
}

.modal-container p {
  margin: 0.5rem 0;
}

.modal-container
.modal-text {
  width: 85%;
  max-width: 90%;
  height: 100%;
  margin: 0;
  overflow-y: auto;
  padding: 0 0.6rem;
}

.modal-container
.modal-text h1 {
  margin: 0.5rem 0 0 0;
  font-size: 1.2rem;
}

.modal-container
.modal-text ul {
  margin: 0.2rem 0 0 0;
  padding-left: 1.5rem;
}

.modal-container
.modal-text li {
  font-size: 1.3rem;
}


.modal-container
.modal-text::-webkit-scrollbar-thumb,
.modal-text::-webkit-scrollbar-track {
  background-color: transparent;
}

.modal-container
.modal-got-it-button {
  width: 50%;
  height: 7dvh;
  border-radius: 10rem;
  border: none;
  color: var(--white);
  background-color: var(--main-green);
  font-weight: bold;
  font-size: 1rem;
  margin: 1rem 0;
}

.modal-container
.modal-got-it-button:hover {
  cursor: pointer;
}

.modal-background {
  width: 100dvw;
  max-width: 100%;
  height: 100dvh;
  max-height: 100%;
  background-color: #000;
  background-color: #000000a3;
  position: absolute;
}


/* Phones */
@media only screen and (max-width: 600px) {
  .category-container
  .category-modal {
    width: 80dvw;
    max-height: 50dvh;
  }

  .category-container
  .category-wrapper h1 {
    font-size: 2rem;
    margin-bottom: 0;
  }

  .category-container
  .category-wrapper h4 {
    font-size: 1.3rem;
  }

  .category-container
  .category-wrapper
  .category-short-desc p {
    font-size: 1rem;
  }

  .category-container
  .category-wrapper
  .category-icon {
    font-size: 8rem;
  }

  .category-container
  .category-wrapper
  .category-buttons button {
    width: 85dvw;
  }

  .modal-container {
    height: 50dvh;
  }

  .modal-container
  .modal-got-it-button {
    height: 8dvh;
  }

  .modal-container
  .modal-text li {
    font-size: 0.9rem;
  }
}

/* Portrait tablets and large phones */
@media only screen and (min-width: 601px) and (max-width: 1024px) {
  .category-container
  .category-modal {
    width: 80dvw;
    max-height: 50dvh;
  }

  .category-container
  .category-wrapper h1 {
    font-size: 3.5rem;
    margin: 0;
  }

  .category-container
  .category-wrapper h4 {
    font-size: 2.5rem;
  }

  .category-container
  .category-wrapper
  .category-short-desc p {
    font-size: 1.8rem;
  }

  .category-container
  .category-wrapper
  .category-icon {
    font-size: 14rem;
  }

  .modal-container p {
    font-size: 1.5rem;
  }

  .modal-container
  .modal-container-top h2 {
    font-size: 2rem;
  }

  .modal-container
  .modal-x-button {
    font-size: 1.5rem;
  }

  .modal-container
  .modal-got-it-button{
    height: 7dvh;
    font-size: 1.5rem;
  }

  .category-container
  .category-wrapper
  .category-buttons
  button {
    font-size: 1.7rem;
    width: 85dvw;
  }
}

/* Small laptops and desktops */
@media only screen and (min-width: 1025px) {

}