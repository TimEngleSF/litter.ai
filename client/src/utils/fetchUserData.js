/* eslint-disable consistent-return */
import Cookies from 'js-cookie';
import fetchData from './fetch';

export async function fetchLogin(body) {
  try {
    const response = await fetchData('login', 'POST', body);
    if (response.token) {
      Cookies.set('authToken', response.token, { expires: 7 });
      Cookies.set('username', response.user.displayUsername, { expires: 7 });
    }
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchLogOut() {
  try {
    await fetchData('logout', 'POST');
    await Cookies.remove('authToken');
    await Cookies.remove('username');
  } catch (err) {
    console.error(err);
  }
}

export async function fetchLeaderboardData(path) {
  try {
    const response = await fetchData(path, 'GET');
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchProfileData(user) {
  try {
    const response = await fetchData(`photo/${user}`, 'GET');
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchRegister(body) {
  try {
    const response = await fetchData('register', 'POST', body);
    if (response.status === 'pending') {
      const loginBody = {
        email: body.email,
        password: body.password,
      };
      const loginResponse = await fetchLogin(loginBody);
      return loginResponse;
    }
    return response;
  } catch (err) {
    console.error(err);
  }
}
