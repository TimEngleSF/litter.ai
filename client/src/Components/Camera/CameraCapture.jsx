import React, { useState } from 'react';
import ImageUploaded from './ImageUploaded';
import AddImage from './AddImage';
import '../../css/Capture.css';

export default function CameraCapture() {
  const [image, setImage] = useState(null);

  return (
    <div className="capture-container main-container">
      {image
        ? <ImageUploaded image={image} setImage={setImage} />
        : <AddImage setImage={setImage} />}
      <script src="ImageCapture.js" />
    </div>
  );
}
