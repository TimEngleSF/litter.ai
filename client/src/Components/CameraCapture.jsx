import React, { useState, useRef } from 'react';
import { Link } from 'react-router-dom';
import { CiCirclePlus } from 'react-icons/ci';
import Dropdown from './Dropdown';
import '../css/Capture.css';

export default function CameraCapture() {
  const [image, setImage] = useState(null);
  const [imageSubmitted, setImageSubmitted] = useState(false);
  const [imageCategory, setImageCategory] = useState('glass');
  const captureRef = useRef(null);

  const openCamera = () => {
    captureRef.current.click();
  };

  const removePhoto = () => {
    setImage(null);
  };

  const handleImage = (e) => {
    const imageData = [e.target.files[0]];
    setImage({
      imagePreview: URL.createObjectURL(new Blob(imageData, { type: 'application/zip' })),
      imageFile: e.target.files[0],
    });
  };

  const handlePhotoSubmit = () => {
    setImageSubmitted(true);
  };

  return (
    <div className="capture-container main-container">
      {/* Image has been uploaded; preview the image */}
      {image
        ? (
          <div className="capture-image-wrapper">
            <div className="capture-image-preview">
              <button
                type="button"
                className="image-x-button"
                onClick={removePhoto}
              >
                &#x2715;
              </button>
              <img alt="" src={image.imagePreview} className="capture-image" />
            </div>
            {/* Image has been uploaded and submitted to the AI */}
            {imageSubmitted
              ? (
                <div className="capture-classify">
                  <h1>Help us classify</h1>
                  <div className="capture-dropdown">
                    <Dropdown
                      setImageCategory={setImageCategory}
                      imageCategory={imageCategory}
                    />
                  </div>
                  <p>
                    Is this&nbsp;
                    {imageCategory}
                    ?
                    If it is not,
                    please select the correct category from the dropdown.
                  </p>
                  <div className="capture-confirm-button lower-buttons">
                    <Link to={`/success/${imageCategory}`}>
                      <button type="button">
                        Confirm
                      </button>
                    </Link>
                  </div>
                </div>
              )
              : (
                <div className="capture-submit-button lower-buttons">
                  <button
                    type="button"
                    onClick={handlePhotoSubmit}
                  >
                    Submit
                  </button>
                </div>
              )}
          </div>
        )
        /* Image has not been uploaded, opens device's camera or gives option to upload an image */
        : (
          <>
            <div className="capture-top-section">
              <CiCirclePlus onClick={openCamera} className="capture-icon" />
              <input
                type="file"
                accept="image/*"
                id="cameraFileInput"
                ref={captureRef}
                capture="environment"
                onChange={handleImage}
                style={{ display: 'none' }}
              />
              <div className="capture-icon-bg" />
              <h1>Take a photo of the item</h1>
              <p>
                <span>Click on the camera to open up your device&#x27;s camera.</span>
                <br />
                <span>Please take the photo in portrait mode for best results.</span>
              </p>
            </div>
            <div className="capture-buttons lower-buttons">
              <Link to="/">
                <button className="button-home" type="button">
                  Home
                </button>
              </Link>
            </div>
          </>
        )}
      <script src="ImageCapture.js" />
    </div>
  );
}
