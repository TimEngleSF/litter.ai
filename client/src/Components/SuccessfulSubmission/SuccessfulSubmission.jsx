/* eslint-disable react/jsx-no-useless-fragment */
import React, { useState, useEffect } from 'react';
import { Link, useParams, useNavigate } from 'react-router-dom';
import { v4 as uuid } from 'uuid';
import SubmissionModal from './SubmissionModal';
import Icons from '../Icons';
import categoryData from '../../MockData/mockCategoryData';
import '../../css/SuccessfulSubmission.css';

export default function SuccessfulSubmission() {
  const { category } = useParams();
  const navigate = useNavigate();
  const [type] = useState(categoryData[category]);
  const [categoryName] = useState(category);
  const [modalOpen, setModalOpen] = useState(false);

  const separateString = (text) => (
    text.split('\n').map((line) => (
      <p key={uuid()}>{line}</p>
    ))
  );

  useEffect(() => {
    if (!categoryData[category]) {
      navigate('/404', { replace: true });
    }
  }, [category, navigate]);

  return (
    <>
      {type
      && (
      <div className="category-container main-container">
        <div className="category-wrapper">
          <h4>{type.category}</h4>
          <Icons name={categoryName} classname="category-icon" />
          <h1>{type.name}</h1>
          <div className="category-short-desc">{separateString(type.description)}</div>
          <div className="category-buttons lower-buttons">
            <button
              type="button"
              data-testid="modal-learn-more"
              onClick={() => { setModalOpen(!modalOpen); }}
            >
              Learn More
            </button>
            <Link to="/capture">
              <button type="button">
                Capture another photo
              </button>
            </Link>
            <Link to="/">
              <button className="button-home" type="button">
                Home
              </button>
            </Link>
          </div>
        </div>
        {modalOpen && (
          <>
            <div className="category-modal">
              <SubmissionModal
                modalOpen={modalOpen}
                setModalOpen={setModalOpen}
                separateString={separateString}
                type={type}
              />
            </div>
            <div
              className="modal-background"
              role="presentation"
              onClick={() => { setModalOpen(!modalOpen); }}
              onKeyDown={() => { setModalOpen(!modalOpen); }}
            />
          </>
        )}
      </div>
      )}
    </>
  );
}
