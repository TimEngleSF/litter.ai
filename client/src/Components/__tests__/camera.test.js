/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import renderer from 'react-test-renderer';
import {
  act,
  render,
  screen,
  fireEvent,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter as Router, Routes, Route } from 'react-router-dom';
import * as fetchUserData from '../../utils/fetchUserData';
import AddImage from '../Camera/AddImage';
import ImageUploaded from '../Camera/ImageUploaded';
import CameraCapture from '../Camera/CameraCapture';
import SuccessfulSubmission from '../SuccessfulSubmission/SuccessfulSubmission';

// SetState function and mock uploaded image
const setImageMock = jest.fn();
const mockFile = new File(['cameraTest'], 'cameraTest.png', { type: 'image/png' });

describe('AddImage component', () => {
  // Function to simulate an setState function
  const addImage = <Router><AddImage setImage={setImageMock} /></Router>;

  // Mock window.URL.createObjectURL in Jest
  window.URL.createObjectURL = jest.fn();
  afterEach(() => { window.URL.createObjectURL.mockReset(); });

  // Creates a snapshot
  test('AddImage matches the current snapshot', () => {
    const tree = renderer.create(addImage).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Uploaded image is saved as a state', async () => {
    render(
      <Router>
        <CameraCapture />
      </Router>,
    );

    // Checking if it is on the camera capture page
    expect(screen.getByText('Take a photo of the item')).toBeInTheDocument();

    // Mocks uploading an image file
    fireEvent.click(screen.getByTestId('camera-plus'));
    const fileUpload = screen.getByTestId('camera-input');
    await userEvent.upload(fileUpload, mockFile);

    // Checks if the image has been uploaded, state has changed, and submit button is visible
    expect(screen.getByTestId('image-preview')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Submit' })).toBeInTheDocument();
  });
});

describe('ImageUploaded component', () => {
  const imageUpload = (
    <Router>
      <ImageUploaded image={mockFile} setImage={setImageMock} />
    </Router>
  );

  // Mock response data from the AI after submitting a photo
  const mockImageData = { class: 'compost', confidence: 0.4749128520488739 };
  const mockCall = async (data) => {
    await jest.spyOn(fetchUserData, 'fetchImageToAI').mockResolvedValue(data);
  };

  beforeEach(() => { mockCall(mockImageData); });
  afterEach(() => { jest.restoreAllMocks(); });

  test('Clicking on X removes the image', async () => {
    render(imageUpload);
    await fireEvent.click(screen.getByRole('button', { name: /image-x-button/i }));
    // SetState function that removes the image from the image state
    expect(setImageMock).toHaveBeenCalled();
  });

  test('Clicking on submit sends the image to the AI', async () => {
    render(imageUpload);
    // Submitting mock image data to mock AI
    await act(() => fireEvent.click(screen.getByRole('button', { name: /Submit/i })));
    expect(screen.getByText(/Help us classify/i)).toBeInTheDocument();
  });

  test("Confirming the image's category routes to the category's information page", async () => {
    render(
      <Router initialEntries={['/capture']}>
        <Routes>
          <Route path="/capture" element={<ImageUploaded image={mockFile} setImage={setImageMock} />} />
          <Route path="/success/:category" element={<SuccessfulSubmission />} />
        </Routes>
      </Router>,
    );
    await act(() => fireEvent.click(screen.getByRole('button', { name: /Submit/i })));
    await act(() => fireEvent.click(screen.getByRole('button', { name: /Confirm/i })));
    expect(screen.getByTestId(/compost-icon/i)).toBeInTheDocument();
  });
});
