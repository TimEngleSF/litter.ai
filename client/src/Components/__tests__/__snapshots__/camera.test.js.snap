// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddImage component AddImage matches the current snapshot 1`] = `
Array [
  <div
    className="capture-top-section"
  >
    <svg
      className="capture-icon"
      data-testid="camera-plus"
      fill="currentColor"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        id="Circle_Plus"
      >
        <g>
          <path
            d="M15,12.5H12.5V15a.5.5,0,0,1-1,0V12.5H9a.5.5,0,0,1,0-1h2.5V9a.5.5,0,0,1,1,0v2.5H15A.5.5,0,0,1,15,12.5Z"
          />
          <path
            d="M12,21.932A9.934,9.934,0,1,1,21.932,12,9.944,9.944,0,0,1,12,21.932ZM12,3.065A8.934,8.934,0,1,0,20.932,12,8.944,8.944,0,0,0,12,3.065Z"
          />
        </g>
      </g>
    </svg>
    <input
      accept="image/*"
      capture="environment"
      data-testid="camera-input"
      id="cameraFileInput"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className="capture-icon-bg"
    />
    <h1>
      Take a photo of the item
    </h1>
    <p>
      <span>
        Click on the icon to open up your device's camera.
      </span>
      <br />
      <span>
        Please take the photo in portrait mode for best results.
      </span>
    </p>
  </div>,
  <div
    className="capture-buttons lower-buttons"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <button
        className="button-home"
        type="button"
      >
        Home
      </button>
    </a>
  </div>,
]
`;
